### PR TITLE
Implemented GameLogger and GameStatistics in GameLoggerImpl

### PR DESCRIPTION
**Pull request by hilalcomak**
I will add the tests before merging.

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
+package ch.uzh.ifi.hase.soprafs22.game.logger;
+
+/**
+ * A Logger for {@link ch.uzh.ifi.hase.soprafs22.game.Game}.
+ */
+public interface GameLogger {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java:6]: can you create a package and namespace "interfaces" inside of logger and move the interfaces there, please? :)

And can you rename the interfaces to a name with a prepending capital I, like IGameLogger, IGameStatistics?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameStatistics.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+import java.util.Map;
+
+/**
+ * Provides statistical information to the client about a given game.
+ */
+public interface GameStatistics {
+    /**
+     * Returns the number of life units per player at each turn.
+     */
+    Map<Long, List<Integer>> unitsPerPlayer();
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameStatistics.java:13]: a Map of a Long and a List<Integer>?

Don't we have something from the PlayerDecorator already at hand to count the units of a player?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+import ch.uzh.ifi.hase.soprafs22.game.logger.GameLogger;
+import ch.uzh.ifi.hase.soprafs22.game.logger.GameStatistics;
+
+import java.util.*;
+
+/**
+ * Implementation of GameLogger to provide GameStatistics.
+ */
+public final class GameLoggerImpl implements GameStatistics, GameLogger {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java:11]: rename to GameLogger, please

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    private final List<Map<Long, Integer>> turnSnapshots = new ArrayList<>();
+    private int totalMoves = 0;
+    private int turn = 0;
+
+    public static Builder newBuilder(){
+        return new Builder();
+    }
+
+    public static final class Builder {
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java:22]: also here, I am not familiar with a nested Builder.
Why don't we use the simple approach with a public constructor of the outer class?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    public float averageUnitsPerTurn() {
+        int allUnitsInAllTurns = turnSnapshots.stream().flatMap(snapshot -> snapshot.values().stream()).mapToInt(Integer::intValue).sum();
+        return (float) allUnitsInAllTurns / (float) (turnSnapshots.size() - 1); // -1 for the ghost entry.
+    }
+
+    private List<Integer> killsPerTurn() {
+        List<Integer> result = new ArrayList<>();
+        for (int i = 1; i < turnSnapshots.size(); i++) {
+            // All the units that were alive in previous turn
+            int prev = turnSnapshots.get(i - 1).values().stream().mapToInt(Integer::intValue).sum();
```
Comment (paolovic) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java:67]: this looks very complicated just for computing the kills per turn in my opinion.
Is our design that "bad", that such a "simple" job turns out to such a line?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    /**
+     * Inform that a unit was killed.
+     * @param turn in which it was killed.
+     * @param playerId id of player that lost the unit.
+     */
+    void unitKilledAtTurn(int turn, long playerId);
```
Comment (apple00juice) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java:13]: We may also want to know which unit it was ? Where it was? What killed it? Who killed?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    /**
+     * Inform that a unit moved.
+     * @param turn in which it moved.
+     */
+    void move(int turn);
```
Comment (apple00juice) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java:19]: Similar here which unit, of whom, where did it move?

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/Game.java [hunk 8]
```diff
@@ -20,9 +23,12 @@ public class Game {
     private final Map<String, PlayerDecorator> playerMap;
     private GameMap gameMap;
     private int turnNumber;
-    private String[] turnOrder;
+    private final String[] turnOrder;
     private boolean running;
 
+    private final GameLogger logger;
```
Comment (apple00juice) [inline src/main/java/ch/uzh/ifi/hase/soprafs22/game/Game.java:29]: Why do we need two ? But I haven't seen this pattern before, so I don't understand it ^^

**Review by apple00juice — COMMENTED** [review]
Similar to Paul, I don't know why we need the Builder pattern. But I just glanced over the builder pattern.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameStatistics.java (hunk 2 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameStatistics.java:13]: Yes and no. This is exactly the information we need to keep track of the stats. PlayerDecorator has much more information (such as the units themselves, and taking a snapshot of that has a much higher memory footprint). 
Moreover, keeping the counter separate, allows for separation of responsabilities (see https://java-design-patterns.com/principles/#single-responsibility-principle).

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java (hunk 1 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java]: afaik, what I did is the standard for interfaces and implementations (see https://stackoverflow.com/questions/2814805/java-interfaces-implementation-naming-convention ). I changed it to match what you say for consistency with other code.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java (hunk 3 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java]: Sure, done.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java (hunk 4 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java]: I just copied the structure of AutoValue.Builder (see https://github.com/google/auto/blob/master/value/userguide/builders.md). Nesting the Builder inside the class allows keeping the constructor private, so the only way to get those objects is through the builder.

I deleted the builder, they are only to make code easier to read and this one seems to be a problem.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java (hunk 5 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/implementation/GameLoggerImpl.java]: It is literally a single line. I have rewritten in more lines, maybe that way it is easier to understand.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java (hunk 6 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java]: I am not sure why we need to know all those things. I collected the minimum information based on the mocks in MileStone2Final.pdf, page 39. If we are targeting other kinds of statistics let me know and I will change it to cover those.

If we are not going to implement any statistics based on that information, it seems like a waste of time and effort.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java (hunk 7 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java]: See previous comment.

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/Game.java (hunk 8 above)
Comment (hilalcomak) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/Game.java]: Indeed, removed one. Thanks for the catch!

File: src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java (hunk 6 above)
Comment (apple00juice) [reply src/main/java/ch/uzh/ifi/hase/soprafs22/game/logger/GameLogger.java]: I see, valid point. If we would need to expand it in the future we might want it. But as we have our requirements we should do just that 👍

**Review by paolovic — APPROVED** [review]

**Pull request merged**